File: diemJava/src/main/java/diem/component/Pacemaker.java
package diem.component;

import diem.model.Block;
import diem.model.QuorumCertificate;
import diem.model.TimeoutCertificate;
import diem.model.TimeoutInfo;
import diem.model.TimeoutMessage;

public class Pacemaker {

  public static int currentRound;
  static TimeoutCertificate lastRoundTc;
  int pendingTimeouts;
  Safety safety = new Safety();

  public static void advanceRound(TimeoutCertificate timeoutCertificate) {
  }

  public static void advanceRound(int round) {
  }

  // TODO
  public int getRoundTimer(int r) {
    return 1; // round timer formulae
  }

  // TODO
  public static void startTimer(int newRound) {
    stopTimer(currentRound);
    currentRound = newRound;
    // Start Local Timer
  }

  private static void stopTimer(int currentRound) {
  }

  public void localTimeoutRound() {
    saveConsensusState();
    TimeoutInfo timeoutInfo = safety.makeTimeout(currentRound, BlockTree.highCommitQC, lastRoundTc);
    TimeoutMessage timeoutMessage = new TimeoutMessage(timeoutInfo, lastRoundTc, BlockTree.highCommitQC);

    //API CALL
    //broadcast TimeoutMessage
  }

  private void saveConsensusState() {
  }

  // TODO
  public static TimeoutCertificate processRemoteTimeout(TimeoutMessage message) {
    TimeoutInfo timeoutInfo = message.getTimeoutInfo();

    if (timeoutInfo.round < currentRound) {
      return null;
    }

    // Pending Timeout Checks
    return null;
  }

  public static boolean advanceRoundTc(TimeoutCertificate tc) {
    if (tc == null || tc.round < currentRound) {
      return false;
    }
    lastRoundTc = tc;
    startTimer(tc.round + 1);
    return true;
  }

  public boolean advanceRoundQc(QuorumCertificate qc) {
    if (qc.voteInfo.round < currentRound) {
      return false;
    }
    lastRoundTc = null;
    startTimer(qc.voteInfo.round + 1);
    return true;
  }
}

File: diemBFT/model/pacemaker.py
from timeout_message import TimeoutMessage
from block_tree import BlockTree


class Pacemaker:
    def __init__(self, current_round, last_round_tc, pending_timeouts):
        self.current_round = current_round
        self.last_round_tc = last_round_tc
        self.pending_timeouts = pending_timeouts


    def get_round_timer(self, r):
        return round_timer_formula

    def star_timer(self, new_round):
        stop_timer(self.current_round)

    def stop_timer(self, current_round):
        pass

    def local_timeout_round(self):
        save_consensus_state()
        timeout_info = Safety.make_timeout(self.current_round, BlockTree.high_qc, last_round_tc)
        broadcast TimeoutMessage(timeout_info, last_round_tc, BlockTree.high_commit_qc)

    def process_remote_timeout(self, tmo):
        tmo_info = tmo.tmo_info

    def advance_round(self, tc):
        if tc is None or tc.round < self.current_round:
            return False
        last_round_tc = tc
        start_timer(tc.round + 1)
        return True

    def advance_round_qc(self, qc):
        if qc.vote_info.round < self.current_round:
            return False
        last_round_tc = None
        start_timer(qc.vote_info.round + 1)
        return True


